buffer.go

```diff
@@ -49,11 +50,13 @@ func (b *buffer) fill(need int) error {
 	// grow buffer if necessary
 	// TODO: let the buffer shrink again at some point
 	//       Maybe keep the org buf slice and swap back?
-	if need > len(b.buf) {
+	if need > cap(b.buf) {
 		// Round up to the next multiple of the default size
 		newBuf := make([]byte, ((need/defaultBufSize)+1)*defaultBufSize)
 		copy(newBuf, b.buf)
 		b.buf = newBuf
+	} else if need > len(b.buf) {
```

Comment:
This could be nested instead. `need > cap(b.buf)` can only be true if also `need > len(b.buf)` is true.

Reply:
If I follow where you're going, it force an unneeded comparison in the common case `len` == `cap`

Is this what you're thinking of?
```go
if need > len(b.buf) {
    if need > cap(b.buf) {
        // create larger buffer
    } else {
       // expand existing buffer
   }
}
```
I personally find less nesting is clearer and simply to reason about.

Reply:
Yes, except for that it makes an even more common case cheaper: `need <= len(b.buf)`